src/Sylius/Bundle/CoreBundle/Controller/OrderController.php

```diff
@@ -24,6 +28,76 @@ class OrderController extends BaseOrderController
      *
      * @return Response
      */
+    public function completeAction(Request $request)
+    {
+        $configuration = $this->requestConfigurationFactory->create($this->metadata, $request);
+
+        $this->isGrantedOr403($configuration, ResourceActions::UPDATE);
+        $order = $this->findOr404($configuration);
+
+        $form = $this->resourceFormFactory->create($configuration, $order);
+
+        if (in_array($request->getMethod(), ['POST', 'PUT', 'PATCH'], true) && $form->handleRequest($request)->isValid()) {
+            $order = $form->getData();
+
+            $event = $this->eventDispatcher->dispatchPreEvent(ResourceActions::UPDATE, $configuration, $order);
+
+            if ($event->isStopped() && !$configuration->isHtmlRequest()) {
+                throw new HttpException($event->getErrorCode(), $event->getMessage());
+            }
+            if ($event->isStopped()) {
+                $this->flashHelper->addFlashFromEvent($configuration, $event);
+
+                return $this->redirectHandler->redirectToResource($configuration, $order);
+            }
+
+            try {
+                if ($configuration->hasStateMachine()) {
+                    $this->stateMachine->apply($configuration, $order);
+                }
+
+                $this->manager->flush();
+            } catch (OptimisticLockException $exception) {
+                $this->addFlash('error', 'sylius.checkout.complete_error');
+
+                return $this->redirectHandler->redirectToRoute($configuration, 'sylius_shop_homepage');
+            }
+
+            $this->eventDispatcher->dispatchPostEvent(ResourceActions::UPDATE, $configuration, $order);
+
+            $this->getOrderEmailManager()->sendConfirmationEmail($order);
```

Comment:
With emails on state machine callbacks I see the following issues:
1. We can't be sure some will not configure the environment to not use Spool and it shouldn't be a requirement for using Sylius
2. State machine calls are made before flush, so every single operation that uses state machine will have to have additional logic for handling errors. There are many exceptions that can occur on `flush` (not only locking), then we check for them and clear the spool in every place that uses state machine...
3. Notifying external system/user should be done only after the state of the application actually changed, so it's natural that it should happen always after the changes are saved in the system, so after the `flush`

I would say we should move it to listeners.